File: agentcash/token.py
"""The budget token: a macaroon-style chained MAC.

    sig_0 = HMAC(root_key, "root_id|root_token_id")
    sig_n = HMAC(sig_{n-1}, caveat_n)

Two properties fall straight out of that construction, and they are the whole
reason this shape was chosen over a JWT:

1. ANYONE HOLDING THE TOKEN CAN ATTENUATE IT, OFFLINE, WITH NO ISSUER CALL.
   The current signature *is* the key for the next link. An agent can carve a
   smaller budget for a sub-agent mid-task, on a plane, with the network down.

2. NOBODY CAN REMOVE A CAVEAT.
   Un-appending requires sig_{n-1}, which requires the root key. Delegation is
   a one-way ratchet: tighter, never looser.

This is the inversion of Bitcoin's key model. A Bitcoin key is unlimited,
eternal, and irrevocable authority in a single bearer secret. That model is
what makes Bitcoin safe for a careful sovereign and catastrophic in the hands
of something that leaks its own context. Here the bearer secret is bounded,
expiring, revocable, and by itself worthless (see CNF below).
"""

import hmac
import hashlib
import secrets as _secrets
from dataclasses import dataclass
from typing import Tuple

from .caveats import ser, de, SUB, CNF


def _mac(key: bytes, msg: str) -> bytes:
    return hmac.new(key, msg.encode("utf-8"), hashlib.sha256).digest()


def new_id(prefix: str) -> str:
    return f"{prefix}_{_secrets.token_hex(8)}"




@dataclass(frozen=True)
class Token:
    root_id: str                 # the principal (who is ultimately liable)
    root_token_id: str           # id of the first grant in this chain
    caveats: Tuple[str, ...]
    sig: str

    # -- structure ----------------------------------------------------------
    @property
    def lineage(self) -> Tuple[str, ...]:
        """Every token id from the root grant down to this one.

        Derived from the SUB caveats, never from a self-declared field: a
        forged lineage would have to survive the signature chain, and it can't.
        """
        out = [self.root_token_id]
        for c in self.caveats:
            k, v = de(c)
            if k == SUB:
                out.append(v)
        return tuple(out)

    @property
    def token_id(self) -> str:
        return self.lineage[-1]

    @property
    def depth(self) -> int:
        return len(self.lineage) - 1

    def scoped(self):
        """[(owner_token_id, kind, value)] - which link in the chain owns each caveat.

        Matters because a `max_total` written by the principal must be counted
        against the *entire subtree's* spending, while one written by a
        sub-agent counts only its own. Without this the ancestor's cap would be
        trivially escaped by delegating.
        """
        owner = self.root_token_id
        out = []
        for c in self.caveats:
            k, v = de(c)
            if k == SUB:
                owner = v
            else:
                out.append((owner, k, v))
        return out

    def find(self, kind):
        return [(o, v) for (o, k, v) in self.scoped() if k == kind]

    # -- crypto -------------------------------------------------------------
    def _extend(self, added) -> "Token":
        s = bytes.fromhex(self.sig)
        for c in added:
            s = _mac(s, c)
        return Token(self.root_id, self.root_token_id, self.caveats + tuple(added), s.hex())

    def attenuate(self, added_caveats) -> "Token":
        """Append caveats without touching the issuer. Offline. Tighter only."""
        return self._extend(list(added_caveats))

    def delegate(self, sub_token_id: str, binding: dict, added_caveats) -> "Token":
        """`binding` is the child's PUBLIC key material. Nothing secret goes in.

        Because it rides inside the signature chain, a verifier needs no prior
        knowledge of the child - which is what makes offline delegation safe.
        """
        chain = [ser(SUB, sub_token_id), ser(CNF, binding)] + list(added_caveats)
        return self._extend(chain)

    # -- wire ---------------------------------------------------------------
    def to_dict(self):
        return {
            "root_id": self.root_id,
            "root_token_id": self.root_token_id,
            "caveats": list(self.caveats),
            "sig": self.sig,
        }

    @staticmethod
    def from_dict(d):
        return Token(d["root_id"], d["root_token_id"], tuple(d["caveats"]), d["sig"])


def mint(root_key: bytes, root_id: str, root_token_id: str,
         binding: dict, caveats) -> Token:
    s = _mac(root_key, f"{root_id}|{root_token_id}")
    t = Token(root_id, root_token_id, tuple(), s.hex())
    return t._extend([ser(CNF, binding)] + list(caveats))


def verify_sig(root_key: bytes, token: Token) -> bool:
    s = _mac(root_key, f"{token.root_id}|{token.root_token_id}")
    for c in token.caveats:
        s = _mac(s, c)
    return hmac.compare_digest(s.hex(), token.sig)


# -- proof of possession -----------------------------------------------------
def canonical_request(token_id, payee, amount, purpose, idem_key) -> str:
    """Everything the proof commits to. Change any field and the proof dies."""
    return f"{token_id}|{payee}|{int(amount)}|{purpose}|{idem_key}"

File: js/src/token.js
// The budget token: a macaroon-style chained MAC.
//
//   sig0 = HMAC(root_key, root_id + "|" + root_token_id)
//   sigN = HMAC(sigN-1, caveatN)
//
// Any holder can append a caveat offline, because the current signature is the
// key for the next link. Nobody can remove one, because that needs the
// previous signature, which needs the root key. Delegation is a ratchet.

import { createHmac, randomBytes } from "node:crypto";
import { ser, de, SUB, CNF } from "./caveats.js";

const mac = (key, msg) => createHmac("sha256", key).update(msg, "utf8").digest();

export const newId = prefix => `${prefix}_${randomBytes(8).toString("hex")}`;

export class Token {
  constructor(rootId, rootTokenId, caveats, sig) {
    this.root_id = rootId;
    this.root_token_id = rootTokenId;
    this.caveats = [...caveats];
    this.sig = sig;
  }

  get lineage() {
    const out = [this.root_token_id];
    for (const c of this.caveats) {
      const [k, v] = de(c);
      if (k === SUB) out.push(v);
    }
    return out;
  }

  get tokenId() { return this.lineage[this.lineage.length - 1]; }
  get depth() { return this.lineage.length - 1; }

  // Which link in the chain owns each caveat. An ancestor's max_total counts
  // its whole subtree; a child's counts only its own.
  scoped() {
    let owner = this.root_token_id;
    const out = [];
    for (const c of this.caveats) {
      const [k, v] = de(c);
      if (k === SUB) owner = v; else out.push([owner, k, v]);
    }
    return out;
  }

  find(kind) { return this.scoped().filter(([, k]) => k === kind).map(([o, , v]) => [o, v]); }

  _extend(added) {
    let s = Buffer.from(this.sig, "hex");
    for (const c of added) s = mac(s, c);
    return new Token(this.root_id, this.root_token_id,
                     [...this.caveats, ...added], s.toString("hex"));
  }

  attenuate(added) { return this._extend(added); }

  delegate(subTokenId, binding, added = []) {
    return this._extend([ser(SUB, subTokenId), ser(CNF, binding), ...added]);
  }

  toJSON() {
    return { root_id: this.root_id, root_token_id: this.root_token_id,
             caveats: this.caveats, sig: this.sig };
  }

  static fromJSON(d) {
    return new Token(d.root_id, d.root_token_id, d.caveats, d.sig);
  }
}

export function mint(rootKey, rootId, rootTokenId, binding, caveats) {
  const s = mac(rootKey, `${rootId}|${rootTokenId}`);
  return new Token(rootId, rootTokenId, [], s.toString("hex"))
    ._extend([ser(CNF, binding), ...caveats]);
}

export function verifySig(rootKey, token) {
  let s = mac(rootKey, `${token.root_id}|${token.root_token_id}`);
  for (const c of token.caveats) s = mac(s, c);
  return s.toString("hex") === token.sig;
}

// Everything the proof commits to. Change any field and the proof dies.
export const canonicalRequest = (tokenId, payee, amount, purpose, idemKey) =>
  `${tokenId}|${payee}|${Math.trunc(amount)}|${purpose}|${idemKey}`;
